### PR TITLE
New BTD class layout and member functions - Part 2

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
To be merged after PR #1085 .

In this PR, the functions to initialize back-transform snapshots (buffers) and prepare fields are implemented.

- A function to "prepare" fields required bu back-transform functor is implemented. In this function. m_cell_centered_data multifab stores cell-centered data over all levels using cell-centered functors for user-requested fields. Then all the data is averaged-down to the coarsest level.

### CODE REVIEW

File: Source/Diagnostics/BTDiagnostics.cpp [hunk 1]
```diff
@@ -139,17 +163,81 @@ BTDiagnostics::InitializeFieldBufferData ( int i_buffer , int lev)
 {
     auto & warpx = WarpX::GetInstance();
     // 1. Lab-frame time for the i^th snapshot
+    m_t_lab[i_buffer] = i_buffer * m_dt_snapshots_lab;
```
Comment (ax3l) [inline Source/Diagnostics/BTDiagnostics.cpp]: Just a thought: [`.at()`-access](https://en.cppreference.com/w/cpp/container/vector/at) is a bit "safer" for later code-restructuring as it throws an exception in performance uncritical code sections such as here, when accessed out of bounds.
```suggestion
    m_t_lab.at(i_buffer) = i_buffer * m_dt_snapshots_lab;
```

File: Source/Diagnostics/BTDiagnostics.cpp [hunk 2]
```diff
@@ -139,17 +163,81 @@ BTDiagnostics::InitializeFieldBufferData ( int i_buffer , int lev)
...
+    // 2. Define domain in boosted frame at level, lev
+    amrex::RealBox diag_dom;
```
Comment (ax3l) [inline Source/Diagnostics/BTDiagnostics.cpp:168]: Maybe worth double-checking for the next min/max loop in moving window scenarios:
the default values for the `RealBox` are `0` and `-1`: https://github.com/AMReX-Codes/amrex/blob/9b2172017f8aa6aea00bda090e7c93e7919ff6eb/Src/Base/AMReX_RealBox.H#L23
Maybe setting it to [`std::numeric_limits`](https://en.cppreference.com/w/cpp/types/numeric_limits/lowest) `max`/`lowest` here avoids any potential corner-cases.

File: Source/Diagnostics/BTDiagnostics.cpp [hunk 3]
```diff
@@ -139,17 +163,81 @@ BTDiagnostics::InitializeFieldBufferData ( int i_buffer , int lev)
...
+    amrex::IntVect hi(1);
+    for (int idim=0; idim < AMREX_SPACEDIM; ++idim) {
+        // lo index with same cell-size as simulation at level, lev.
+        lo[idim] = std::max( static_cast<int>( floor (
+                      ( diag_dom.lo(idim) - warpx.Geom(lev).ProbLo(idim)) /
+                        warpx.Geom(lev).CellSize(idim)) ), 0 );
+        // hi index with same cell-size as simulation at level, lev.
+        hi[idim] = std::max( static_cast<int> ( ceil (
+                      ( diag_dom.hi(idim) - warpx.Geom(lev).ProbLo(idim)) /
+                        warpx.Geom(lev).CellSize(idim) ) ), 0) - 1 ;
```
Comment (ax3l) [inline Source/Diagnostics/BTDiagnostics.cpp]: That was a bit hard to read for me :)
```suggestion
        lo[idim] = std::max(
            0,
            static_cast<int>( floor(
                ( diag_dom.lo(idim) - warpx.Geom(lev).ProbLo(idim) ) /
                  warpx.Geom(lev).CellSize(idim)
            ))
        );
        // hi index with same cell-size as simulation at level, lev.
        hi[idim] = std::max(
            0,
            static_cast<int>( ceil(
                ( diag_dom.hi(idim) - warpx.Geom(lev).ProbLo(idim) ) /
                  warpx.Geom(lev).CellSize(idim)
            ))
        ) - 1;
```

One could make the 2nd argument potentially a named, temporary, const variable :)

File: Source/Diagnostics/BTDiagnostics.cpp [hunk 4]
```diff
@@ -139,17 +163,81 @@ BTDiagnostics::InitializeFieldBufferData ( int i_buffer , int lev)
...
+        diag_dom.setLo(idim, std::max(m_lo[idim],warpx.Geom(lev).ProbLo(idim)) );
+        diag_dom.setHi(idim, std::min(m_hi[idim],warpx.Geom(lev).ProbHi(idim)) );
+    }
     // 2. Define domain in boosted frame at level, lev
     // 3. Initializing the m_buffer_box for the i^th snapshot.
     //    At initialization, the Box has the same index space as the boosted-frame
     //    As time-progresses, the z-dimension indices will be modified based on
     //    current_z_lab
+    amrex::IntVect lo(0);
+    amrex::IntVect hi(1);
```
Comment (ax3l) [inline Source/Diagnostics/BTDiagnostics.cpp]: Looking at the logic below, I think a better default value for hi would be:
```suggestion
    amrex::IntVect hi(-1);
```

File: Source/Diagnostics/BTDiagnostics.cpp [hunk 5]
```diff
@@ -161,6 +249,12 @@ BTDiagnostics::DefineCellCenteredMultiFab(int lev)
     // This MultiFab will store all the user-requested fields in the boosted-frame
     auto & warpx = WarpX::GetInstance();
     // The BoxArray is coarsened based on the user-defined coarsening ratio
+    amrex::BoxArray ba = warpx.boxArray(lev);
```
Comment (ax3l) [inline Source/Diagnostics/BTDiagnostics.cpp:278]: Do we need to manipulate this as reference?

File: Source/Diagnostics/BTDiagnostics.cpp (hunk 2 above)
Comment (ax3l) [reply Source/Diagnostics/BTDiagnostics.cpp:168]: Offline discussed: Ah no, this is not a `x = min(x,y);` pattern - sorry, overlooked :)

Comment (RevathiJambunathan) [reply Source/Diagnostics/BTDiagnostics.cpp:168]: Here we set the lo and hi and not find the min/max by comparing with previous idim. As discussed offline, I will add comments here to better explain this.

File: Source/Diagnostics/BTDiagnostics.cpp (hunk 5 above)
Comment (atmyers) [reply Source/Diagnostics/BTDiagnostics.cpp:278]: We are modifying this `ba` below so we don't want it to be a reference.

`BoxArray` is actually pretty cheap to copy bc it doesn't do a deep copy. The reason is that if you modify a BoxArray in a simple way, e.g. by coarsening, you don't want to create a whole new dense representation, you just remember "I'm like the one, but coarsened by some factor". So I think this is actually an efficient way to do this.

**Review by atmyers — APPROVED** [review]

File: Source/Diagnostics/BTDiagnostics.cpp (hunk 5 above)
Comment (ax3l) [reply Source/Diagnostics/BTDiagnostics.cpp:278]: Thank you for the details! :)

**Review by ax3l — APPROVED** [review]

**Comment by MaxThevenet** [comment]
:tada:

**Pull request merged**